core/src/docs/upgrade.md

```diff
@@ -1,3 +1,7 @@
+# Upgrade to v0.39
+
+There are no public API changes.
```

Comment:
There are some raw API changes, please record them with your own words.

In PR https://github.com/apache/incubator-opendal/pull/2666, we changed the struct layout of `RetryLayer` to support user defined `RetryInterceptor`.

And in PR https://github.com/apache/incubator-opendal/pull/2687, we changed the behavior when `role_arn` has been specified. We used to override `role_arn` simply but since this version, we will make sure to use `asseume_role` with specified `role_arn`.

In PR https://github.com/apache/incubator-opendal/pull/2698, we re-org the internal structure of `opendal::raw::oio` and changed some APIs name.

---

And please double check if there are something missed, thanks!

Reply:
Maybe we should maintain the `upgrade.md` while sending PRs instead in release.

Reply:
Absolutely, the release PR author may not have some context to check these API changes.

Reply:
Would you mind helping me to complete these raw API change log entries?  I have seen all these PR author is you, Maybe you can do it better than me.

Reply:
> Would you mind helping me to complete these raw API change log entries?

Of course!

Reply:
Hi @Xuanwo Waiting for your commit to `update.md` doc in the current PR